### PR TITLE
docs(0004): make plugin.get{Prefix,Account,Info} synchronous

### PR DESCRIPTION
**Pull request by sentientwaffle**
Fixes https://github.com/interledger/rfcs/issues/141

### CODE REVIEW

File: 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ This spec depends on the [ILP spec](../0003-interledger-protocol/).
 | | [**connect**](#connect) ( options ) `⇒ Promise.<null>` |
 | | [**disconnect**](#disconnect) ( ) `⇒ Promise.<null>` |
 | | [**isConnected**](#isconnected) ( ) `⇒ Boolean` |
-| | [**getInfo**](#getinfo) ( ) <code>⇒ Promise.&lt;[LedgerInfo](#class-ledgerinfo)></code> |
+| | [**getInfo**](#getinfo) ( ) <code>⇒ [LedgerInfo](#class-ledgerinfo)</code> |
 | | [**getPrefix**](#getprefix) ( ) `⇒ Promise<String>` |
 | | [**getAccount**](#getaccount) ( ) `⇒ Promise<String>` |
```
Comment (emschwartz) [inline 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md]: What about `getPrefix` and `getAccount`? I think those are in the same boat as `getInfo`

Comment (sentientwaffle) [inline 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md]: Updated

**Comment by justmoon** [comment]
Should these just be fields of `LedgerInfo`?

Thinking out loud: The prefix definitely seems like a ledger property nowadays, so I think that could move to LedgerInfo. The account... less so - it's a property of the plugin, not really of the ledger. So maybe that one we keep as a separate method?

Also we may want to add an event if the ledger info changes, e.g. `info_change`. (If we go with the suggestion above: I don't think the account would ever change (?), unless the prefix changes, which would trigger the `info_change` event, so I don't think we need an event for that.)

**Comment by sharafian** [comment]
Do we want to roll all of this functionality into `getInfo`? It seems like it's becoming a catch-all (a cache-all, even) for anything information the plugin is able to store.

It seems like what usually happens is a component like core will call `getInfo`, and then cache the metadata itself. If this were all synchronous, I think we could split `getInfo` into more methods, which could be called more freely, and wouldn't need to minimize calls by wrapping all this info into one object.

**Comment by sharafian** [comment]
I also think that with these changes, we should bring back `getConnectors`, which could synchronously return a list of connector names (right now, everything that gets connectors runs `getInfo`, and then maps the connectors field to a list of connector names)

**Comment by sentientwaffle** [comment]
@justmoon: I think adding `prefix` to `LedgerInfo` is a good idea (and removing `getPrefix`). Re: `info_change`: would that be an event emitted from the ledger to the plugin?

@sharafian: I'm not in favor of adding `getConnectors`; it adds to every plugin's interface without adding any new behavior. If we add `getConnectors`, why not `getCurrencyCode`, `getPrecision`, etc... having everything rolled into one function seems simpler to me.

**Comment by emschwartz** [comment]
If the calls are going to be synchronous either way, I don't think it matters much whether we have one `getInfo` function that returns a bunch of data or individual functions for the various fields. If there's an update event that seems like an argument for rolling the different fields into one object.

Regarding `getConnectors`, I also don't care whether the data is in the info object or exposed through a separate method but I do like the suggestion of making it an array of connector account names -- unless we can think of some compelling use for including other information about the connectors.

**Comment by sentientwaffle** [comment]
So, info would become:

```
{
  "prefix": "example.red."
  "precision": 10,
  "scale": 4,
  "currencyCode": "USD",
  "currencySymbol": "$",
  "connectors": [ "chloe" ]
}
```

I don't see a reason for `connectors` to be an Object anymore.

**Comment by justmoon** [comment]
Why is `connectors` a list of names instead of a list of ILP addresses?

Seems like we should either add the ledger prefix everywhere we talk about accounts or strip it out everywhere.

**Comment by sentientwaffle** [comment]
> If the routing table entry has a liquidity curve, the connector can respond immediately to the quote request. The connector calculates the amount by combining its rate (to the ledger a.b.x) with the curve it got from the next connector (a.b.x.z).

That's not exactly how it works now (see https://github.com/interledgerjs/ilp-core/blob/35d02b7a2c08da4add4c60beacd1dcacb63352b9/src/lib/core.js#L128-L131). It doesn't check for a liquidity curve, it checks if it is a local or a direct route. Even indirect routes have a curve so that we can recompute the source amount before relaying a quote (https://github.com/interledgerjs/ilp-core/blob/35d02b7a2c08da4add4c60beacd1dcacb63352b9/src/lib/core.js#L133-L146)

**Comment by justmoon** [comment]
> Re: info_change: would that be an event emitted from the ledger to the plugin?

Depends on the ledger, could be push or could be polling. Some plugins may not implement this event at all. But having this event (and supporting it in the connector, as well as in the receiver) seems quite useful so that you avoid downtime if a ledger's prefix needs to be changed for any reason.

One more thing: Could you add something to the `connect` method description to the effect of "The plugin should ensure that the information returned by `getInfo` and `getAccount` is available and cached before emitting the connected event."

**Pull request merged**